File: datastructures/strings/python/algorithms/add_binary.py
def addBinary(a, b):
    result = []
    carry = 0

    i = len(a) - 1
    j = len(b) - 1

    while i >= 0 or j >= 0:
        sum = carry

        if i >= 0:
            sum += int(a[i])
        if j >= 0:
            sum += int(b[j])

        result.append(str(sum % 2))
        carry = sum / 2

        i -= 1
        j -= 1

    if carry == 1:
        result.append(str(carry))

    result.reverse()

    return ''.join(result)


def main():
    a = "1010"
    b = "1011"

    print(addBinary(a, b))


if __name__ == "__main__":
    main()

File: datastructures/strings/java/algorithms/AddBinary.java
public class AddBinary {

    public static String addBinary(String a, String b) {
        StringBuilder result = new StringBuilder();

        int carry = 0;
        int i = a.length() - 1;
        int j = b.length() - 1;

        while (i >= 0 || j >= 0) {
            int sum = carry;
            if (i >= 0) {
                sum += a.charAt(i) - '0'; // transforms a char into a number
            }
            if (j >= 0) {
                sum += b.charAt(j) - '0';
            }
            result.append(sum % 2);
            carry = sum / 2;

            i--;
            j--;
        }

        if (carry == 1) {
            result.append(carry);
        }

        return result.reverse().toString();
    }

    public static void main(String[] args) {
        String a = "1010";
        String b = "1011";

        System.out.println(addBinary(a, b));
    }

}
